### ISSUE TITLE
Issues with showing changes in a reverse proxy setup (Git remote address does not match Gerrit address)

### ISSUE THREAD

**Issue by Flaw** [uwolfer/gerrit-intellij-plugin#194]
Our scenario is the following:
Gerrit is listening for ssh connections on ssh://git.example.invalid:29418, but the Gerrit WebUI is exposed through a reverse proxy at http://dev.example.invalid/gerrit. 
In this setup, I can't figure out a way of fetching only changes from the current project.

I can't add the actual Gerrit host as the review server in the plugin settings since I cannot directly access the port where the WebUI is actually listening on git.example.invalid.
If I add a remote to the git repo that matches the host where the WebUI is accessible (dev.example.invalid) I am able to see the proper changes, but fetching commits will not work as the plugin will try to fetch them from the fake remote.

As a side note, everything works fine if I enable fetching of all commits regardless of project, but this will give me a rather silly list of changes that I need to filter using the search box.

Any ideas how this should be configured?

**Comment by uwolfer** [comment]
@Flaw: IMHO adding a remote which matches the reverse proxy address is the right way. I think you should add a remote with a different name and just leave the default remote as-is. That way Git still uses the default remote for fetching, but the plugin is able to match the other address. If this still does not work, please post your .git/config (anonymized).

**Comment by Flaw** [comment]
Hmm, you are indeed correct. I can't get it to fail now no matter how much I try. Granted last time I tried this I was still on IDEA14.0 with probably an older version of the plugin, perhaps it has since been fixed? Or perhaps my configuration has since changed in some crucial manner.

Last time I was looking into this issue it seemed quite random which remote the plugin would use for the fetch when clicking on different changes in the change list, roughly every second click would produce a fetch error in the version control logs. I even forked the plugin with the intention of fixing this but never found the time.

This can be closed from my side, I'll reopen if we manage to reproduce this again.

**Comment by uwolfer** [comment]
@Flaw: Great, thanks for your feedback.

I do not think that anything in this area has recently changed.

**Comment by Flaw** [comment]
This occurred again, it would seem the issue still exist. IDEA 15.0.2 with Gerrit plugin 0.9.7.6-142.
Gerrit is at https://dev.example.com/gerrit, and that's configured in the Gerrit plugin options.
git remote -v:
```
gerrit  https://dev.example.com/gerrit/project (fetch)
gerrit  https://dev.example.com/gerrit/project (push)
origin  ssh://dev@git.example.com:29418/project (fetch)
origin  ssh://dev@git.example.com:29418/project (push)
```

if I do a "Checkout" on any changes I see in the Gerrit dialog, I get a bunch of "Fetch failed" popups, and the following in the VCS console:
```
10:52:28.878: cd C:\workspaces\project
10:52:28.878: git -c core.quotepath=false fetch gerrit refs/changes/63/7063/6 --progress --prune
fatal: unable to access 'https://dev.example.com/gerrit/project/'
```

And the .git/config you asked for:
``` ini
[core]
	repositoryformatversion = 0
	filemode = false
	bare = false
	logallrefupdates = true
	symlinks = false
	ignorecase = true
	hideDotFiles = dotGitOnly
	autocrlf = true
	safecrlf = false
[remote "origin"]
	url = ssh://dev@git.example.com:29418/project
	fetch = +refs/heads/*:refs/remotes/origin/*
[branch "master"]
	remote = origin
	merge = refs/heads/master
[branch "develop"]
	remote = origin
	merge = refs/heads/develop
[gui]
	wmstate = normal
	geometry = 887x427+78+108 171 192
[merge]
	renamelimit = 3000
[remote "sandbox"]
	url = https://git.mirror.com/repos/project_sandbox.git
	fetch = +refs/heads/*:refs/remotes/sandbox/*
[remote "gerrit"]
	url = https://dev.example.com/gerrit/project
	fetch = +refs/heads/*:refs/remotes/gerrit/*
```
I'm not entirely sure what changed. This occurred after I restarted IDEA for the latest Gerrit-plugin update, but I also had just changed passwords meaning I got a access denied when it tried to fetch the pending reviews after the restart. After updating the password in the settings and refreshing the pending reviews list, this situation is what I was left with, If that gives you any ideas.
And as before, it seems to work intermittently, as can be seen from the log:
```
10:52:28.878: cd C:\workspaces\project
10:52:28.878: git -c core.quotepath=false fetch gerrit refs/changes/63/7063/6 --progress --prune
fatal: unable to access 'https://dev.example.com/gerrit/project/': SSL certificate problem: self signed certificate in certificate chain
10:52:29.379: cd C:\workspaces\project
10:52:29.379: git -c core.quotepath=false checkout -b review/developer_a/feature_b FETCH_HEAD
fatal: Cannot update paths and switch to branch 'review/developer_a/feature_b' at the same time.
Did you intend to checkout 'FETCH_HEAD' which can not be resolved as commit?
11:17:59.639: cd C:\workspaces\project
11:17:59.639: git -c core.quotepath=false fetch origin refs/changes/51/7851/1 --progress --prune
remote: Counting objects: 133, done        
remote: Finding sources: 100% (68/68)        
remote: Total 68 (delta 47), reused 68 (delta 47)        
From ssh://git.example.com:29418/project
 * branch            refs/changes/51/7851/1 -> FETCH_HEAD
11:18:13.099: cd C:\workspaces\project
11:18:13.099: git -c core.quotepath=false fetch origin refs/changes/36/7736/3 --progress --prune
remote: Counting objects: 31, done        
remote: Finding sources: 100% (16/16)        
remote: Total 16 (delta 8), reused 16 (delta 8)        
From ssh://git.example.com:29418/project
 * branch            refs/changes/36/7736/3 -> FETCH_HEAD
11:18:23.005: cd C:\workspaces\project
11:18:23.005: git -c core.quotepath=false fetch origin refs/changes/46/5946/5 --progress --prune
From ssh://git.example.com:29418/project
 * branch            refs/changes/46/5946/5 -> FETCH_HEAD
11:18:28.292: cd C:\workspaces\project
11:18:28.292: git -c core.quotepath=false fetch origin refs/changes/63/7063/6 --progress --prune
11:18:29.803: cd C:\workspaces\project
11:18:29.803: git -c core.quotepath=false fetch origin refs/changes/63/7063/6 --progress --prune
remote: Counting objects: 22, done        
remote: Finding sources: 100% (12/12)        
remote: Total 12 (delta 2), reused 10 (delta 2)        
From ssh://git.example.com:29418/project
 * branch            refs/changes/63/7063/6 -> FETCH_HEAD
remote: Counting objects: 22, done        
remote: Finding sources: 100% (12/12)        
remote: Total 12 (delta 2), reused 10 (delta 2)        
From ssh://git.example.com:29418/project
 * branch            refs/changes/63/7063/6 -> FETCH_HEAD
11:18:36.806: cd C:\workspaces\project
11:18:36.806: git -c core.quotepath=false checkout -b review/developer_a/feature_b FETCH_HEAD
Switched to a new branch 'review/developer_a/feature_b'
11:18:37.097: cd C:\workspaces\project
11:18:37.097: git -c core.quotepath=false checkout review/developer_a/feature_b
Already on 'review/developer_a/feature_b'
11:18:52.903: cd C:\workspaces\project
11:18:52.903: git -c core.quotepath=false rm --ignore-unmatch -- .idea/libraries/Maven__com_google_inject_guice_4_0.xml .idea/libraries/Maven__com_google_guava_guava_16_0_1.xml
11:18:56.601: cd C:\workspaces\project
11:18:56.601: git -c core.quotepath=false fetch gerrit refs/changes/58/7558/9 --progress --prune
fatal: unable to access 'https://dev.example.com/gerrit/project/': SSL certificate problem: self signed certificate in certificate chain
fatal: bad object 79f7dabad70c65d53c2915844ed16c7b6d753569
```
That's just me clicking around in the change list, with one test checkout thrown in. I continued messing about the change list for a good 3 minutes or so, and after it started failing it failed consistently.
Let me know if you need something, I'll update here if I figure out a way of reliably reproducing this.

**Comment by uwolfer** [comment]
@Flaw: Does the URL `https://dev.example.com/gerrit/project/` work for git operations or is it just an alias for making project detection work? It looks like this git remote is used for remote operations.

I don't think anything related to this has changed in the last few releases.

**Comment by Flaw** [comment]
It's only an alias for the project detection, as instructed. No git operations will work on that.
The question here is why is it using the gerrit remote. This is what I meant in the original post that simply adding a remote with the proper URL is not a solution as that will break fetching.
What's interesting is why does it sometimes use the origin remote and other times the gerrit remote.

**Comment by uwolfer** [comment]
I have just tried to debug your situation. Git remote operations are done on the remote which matches your Gerrit address. This is required for most people since there are setups with multiple remotes, but Gerrit review commits can only be fetched from the Gerrit remote.

One workaround which should work for you: in settings,

**Comment by uwolfer** [comment]
This should be fixed with PR #324. Please reopen if it does not fix your situation after the next release of this plugin.
